[Opt Overlap] Implement as_functional_optim and create_functional_optim

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71620
* #71608
* #71607
* #71606
* #71605
* __->__ #71604
* #71603
* #71602

Implement 2 helper functions:
- as_functional_optim which takes in a torch.optim class type and arguments and
  creates the corresponding functional optimizer.
- create_functional_optim which takes in the functional optimizer class type
  and constructs it. Note that as_functional_optim calls into
  create_functional_optim.

  The first will be used in future PRs as described in
  https://github.com/pytorch/pytorch/issues/67570 to create a functional
  optimizer from a traditional optimizer. The latter is used in
  _OptimizerHookState to create a functional optimizer.

  Both new helper functions are covered by unittests.

Differential Revision: [D33688995](https://our.internmc.facebook.com/intern/diff/D33688995/)

rohan-varma committed:
[Opt Overlap] Implement as_functional_optim and create_functional_optim

Implement 2 helper functions:
- as_functional_optim which takes in a torch.optim class type and arguments and
  creates the corresponding functional optimizer.
- create_functional_optim which takes in the functional optimizer class type
  and constructs it. Note that as_functional_optim calls into
  create_functional_optim.

  The first will be used in future PRs as described in
  https://github.com/pytorch/pytorch/issues/67570 to create a functional
  optimizer from a traditional optimizer. The latter is used in
  _OptimizerHookState to create a functional optimizer.

  Both new helper functions are covered by unittests.

Differential Revision: [D33688995](https://our.internmc.facebook.com/intern/diff/D33688995/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Implement as_functional_optim and create_functional_optim"


Implement 2 helper functions:
- as_functional_optim which takes in a torch.optim class type and arguments and
  creates the corresponding functional optimizer.
- create_functional_optim which takes in the functional optimizer class type
  and constructs it. Note that as_functional_optim calls into
  create_functional_optim.

  The first will be used in future PRs as described in
  https://github.com/pytorch/pytorch/issues/67570 to create a functional
  optimizer from a traditional optimizer. The latter is used in
  _OptimizerHookState to create a functional optimizer.

  Both new helper functions are covered by unittests.

Differential Revision: [D33688995](https://our.internmc.facebook.com/intern/diff/D33688995/)

[ghstack-poisoned]

cbalioglu commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -16,11 +17,10 @@ class _OptimizerHookState(object):
     ):
-        self.functional_optimizer = functional_optim_cls(
-            [],
+        self.functional_optimizer = create_functional_optim(
```
Now that we have a standalone `create_functional_optim` function, I think this `__init__` has become even more redundant. I think it would be cleaner if we simply replaced it with `from_functional_optim`, meaning changing it to `__init__(self, functional_optim)` that contains the current logic of `from_function_optim`.

rohan-varma replied:
Yep, this is done in the top most diff.

cbalioglu commented on torch/distributed/optim/__init__.py:
```diff
@@ -15,21 +16,12 @@
 from .functional_rprop import _FunctionalRprop
 from .functional_adamax import _FunctionalAdamax
+from .utils import (
+    functional_optim_map,
```
Is there a particular reason why we treat `functional_optim_map` as a public variable? Is it for allowing users to extend it with other optim types?

rohan-varma replied:
I think we should actually hide it, as if we wanted to support extension with other optim types there is a better approach for the developer (using the _OverlappedOptimizer interface built out in the next diff), and we're not planning on end users extending this map at least in the prototype version of this future.

To avoid issues with rebasing, I just went ahead and removed this in the top most diff.

cbalioglu commented on torch/distributed/optim/utils.py:
```diff
@@ -0,0 +1,41 @@
+        [],
+        *args,
+        **kwargs,
+        _allow_empty_param_list=True,
```
Same comment as in #71602. I think `_allow_empty_param_list` condition should be handled inside `_OptimizerHookState`.

rohan-varma replied:
The main issue is that `_OptimizerHookState` will expect a fully constructed functional optimizer, so the caller of that needs to have constructed a functional optimizer which requires passing in _allow_empty_param_list with the rest of the arguments. `create_functional_optim` / `as_functional_optim` eventually becomes quite hidden as we only call it in `_OverlappedOptimizer` which is further called from DDP. 

If we want to specify allow_empty_param_list inside ``_OptimizerHookState`, there are 2 options I can think of:

1. Construct the entire functional optimizer inside `_OptimizerHookState` 
2. Just omit `_allow_empty_param_list` arg here, and then manually set it in `_OptimizerHookState`. This hides the arg a bit better but seems hackier as a developer will have to look in 2 places to see how exactly the functional optimizer is being constructed.

cbalioglu reviewed (commented):
Again, just a few nit questions before approval. Looks good to me.

rohan-varma committed:
Update on "[Opt Overlap] Implement as_functional_optim and create_functional_optim"


Implement 2 helper functions:
- as_functional_optim which takes in a torch.optim class type and arguments and
  creates the corresponding functional optimizer.
- create_functional_optim which takes in the functional optimizer class type
  and constructs it. Note that as_functional_optim calls into
  create_functional_optim.

  The first will be used in future PRs as described in
  https://github.com/pytorch/pytorch/issues/67570 to create a functional
  optimizer from a traditional optimizer. The latter is used in
  _OptimizerHookState to create a functional optimizer.

  Both new helper functions are covered by unittests.

Differential Revision: [D33688995](https://our.internmc.facebook.com/intern/diff/D33688995/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Implement as_functional_optim and create_functional_optim"


Implement 2 helper functions:
- as_functional_optim which takes in a torch.optim class type and arguments and
  creates the corresponding functional optimizer.
- create_functional_optim which takes in the functional optimizer class type
  and constructs it. Note that as_functional_optim calls into
  create_functional_optim.

  The first will be used in future PRs as described in
  https://github.com/pytorch/pytorch/issues/67570 to create a functional
  optimizer from a traditional optimizer. The latter is used in
  _OptimizerHookState to create a functional optimizer.

  Both new helper functions are covered by unittests.

Differential Revision: [D33688995](https://our.internmc.facebook.com/intern/diff/D33688995/)

[ghstack-poisoned]
